### PR TITLE
dialects: Add basic shape inference to stencil.

### PR DESCRIPTION
**Pull request by PapyChacal**
This PR adds basic shape inference to the stencil dialect, and error in case of out-of-bound.

### CODE REVIEW

File: xdsl/transforms/experimental/ConvertStencilToLLMLIR.py [hunk 1]
```diff
@@ -112,12 +121,35 @@ class LoadOpToMemref(RewritePattern):
...
     @op_type_rewrite_pattern
     def match_and_rewrite(self, op: LoadOp, rewriter: PatternRewriter, /):
-        assert isinstance(op.field.owner, Operation)
+        cast = op.field.owner
+        assert isinstance(cast, CastOp)
 
-        rewriter.replace_matched_op([], list(op.field.owner.results))
+        if IndexAttr.min(cast.lb, op.lb) != cast.lb:
+            raise VerifyException(
+                f"Out of bound stencil! The stencil computation requires a field with lower bound at least {op.lb}, got {cast.lb}"
```
[the remark this answers is not in the document]
Comment (georgebisbas) [reply xdsl/transforms/experimental/ConvertStencilToLLMLIR.py:129]: >f"Out of bound stencil!

Can drop this

File: xdsl/dialects/experimental/stencil.py [hunk 2]
```diff
@@ -138,6 +138,43 @@ def size_from_bounds(lb: IndexAttr, ub: IndexAttr) -> Sequence[int]:
             for lb, ub in zip(lb.array.data, ub.array.data)
         ]
 
+    def __neg__(self) -> IndexAttr:
```
[the remark this answers is not in the document]
Comment (PapyChacal) [reply xdsl/dialects/experimental/stencil.py:141]: Adding a TODO as "Needs to be clarified to get out of eperimental"

File: xdsl/transforms/experimental/ConvertStencilToLLMLIR.py [hunk 3]
```diff
@@ -70,6 +71,14 @@ class StoreOpCleanup(RewritePattern):
 
     @op_type_rewrite_pattern
     def match_and_rewrite(self, op: StoreOp, rewriter: PatternRewriter, /):
+
+        owner = op.temp.owner
+
+        assert isinstance(owner, ApplyOp | LoadOp)
+
+        owner.attributes['lb'] = IndexAttr.min(op.lb, owner.lb)
```
[the remark this answers is not in the document]
Comment (PapyChacal) [reply xdsl/transforms/experimental/ConvertStencilToLLMLIR.py:79]: On the attributes setters? :) Lovely!

[the remark this answers is not in the document]
Comment (math-fehr) [reply xdsl/transforms/experimental/ConvertStencilToLLMLIR.py:79]: Now here: https://github.com/xdslproject/xdsl/pull/583 ;)

**Pull request merged**